### ISSUE TITLE
Volume keyframes not copied in a copy - paste action

### ISSUE THREAD

**Issue by jep-fa** [jliljebl/flowblade#867]
Flowblade 2.4.0 developpement environnement
MLT 6.21.0
GTK 3.24.18
Ubuntu 20.04

Operations :

- select a clip with volume and video  keyframes curves
- Menu Copy
- Menu Paste

Brightness is the same as the original ; the volume is 100.
I don't understand why because the two filters are well cloned.

**Comment by jep-fa** [comment]
I think I have found the reason. The volume is a MultipartFilterObject with several filters. The sequence.clone_clip_and_filters() calls the function mltfilters.clone_filter_object() which clones only one filter.
Can somebody give me information to create a  function clone_multpartfilter_object() and how to distinguish filter and multipartfilter ?

**Comment by jliljebl** [comment]
Oh yeah, we can get rid of the MultipartFilterObject filters in the next cycle

It is too late for this release but the MultipartFilterObject are no longer needed, this is a really old workaround for MLT limtations at the time, MultipartFilterObject was crated to have key framed volume before volume filter supported keyframes.

We will need to keep it around for backwards compatibility, but we will create new filter that will work with copy pasting kayframes.

**Comment by jep-fa** [comment]
Yes, Brightness has a simpler syntax. I tried this in sequence.py (in the else condition below) but the volume is not cloned.

    def clone_clip_and_filters(self, clip, clone_clip):
        
        clone_clip.clip_in = clip.clip_in
        clone_clip.clip_out = clip.clip_out
        clone_clip.filters = []
        
        for f in clip.filters:
            if isinstance(f, mltfilters.FilterObject):
                clone_filter = mltfilters.clone_filter_object(f, self.profile)
                clone_clip.attach(clone_filter.mlt_filter)
                clone_clip.filters.append(clone_filter)
                
            else:  #f is mltfilters.MultiFilterObject
                clone_filter = mltfilters.clone_multipart_filter_object(f, self.profile, clip)
                clone_filter.attach_all_mlt_filters(clone_clip)
                clone_clip.filters.append(clone_filter)
                
       
        return clone_clip

**Comment by jep-fa** [comment]
Now it seems working with an update of filters dictionnary. 
The volume filters of the copied-pasted clip do not appear in the xml export but a reload of the project file shows the  pasted clip with its volume filter (?).

```
def clone_clip_and_filters(self, clip, clone_clip):
    
    clone_clip.clip_in = clip.clip_in
    clone_clip.clip_out = clip.clip_out
    clone_clip.filters = []
    
    for f in clip.filters:
        if isinstance(f, mltfilters.FilterObject):
            clone_filter = mltfilters.clone_filter_object(f, self.profile)
            clone_clip.attach(clone_filter.mlt_filter)
            clone_clip.filters.append(clone_filter)
            
        else:  #f is mltfilters.MultiFilterObject
            clone_filter = mltfilters.clone_multipart_filter_object(f, self.profile, clip)
            # Update the dictionnary
            clone_filter.__dict__.update(f.__dict__)
            clone_filter.attach_all_mlt_filters(clone_clip)
            clone_clip.filters.append(clone_filter)
            
   
    return clone_clip

```

**Comment by jep-fa** [comment]
If this solution is confirmed, the bug in #814 "When cutting a clip, filters info is deleted from the newly created second clip" is fixed.

**Comment by jliljebl** [comment]
So, the appropriate steps with this are in my view: 
  * next we do 2.6 release.  We do not touch any code related to this before release, this is how one opens a door for regressions by doing changes in very late cycle.
  * replace  all *mltfilters.MultiFilterObject* filters with *mltfilters.FilterObject* filters
  * test and make sure that this fixes this issue
  * we mark mltfilters.MultiFilterObject *DEPRECATED*
  * we do *not* add any new code to fix a deprecated object, users can change to using new filters if they wish
  * we look at  #814 separately in its own issue. It might be quite easy to change this for just cut edit actions, let's see about that in next cycle.

> The volume filters of the copied-pasted clip do not appear in the xml export but a reload of the project file shows the pasted clip with its volume filter (?).

This means that MLT C data structures are different from Flowblade Python side data, we absolutely cannot do that. Fixing #814 does not need fixes in *clone_clip_and_filters()* once we get rid of *mltfilters.MultiFilterObject*.

**Comment by jep-fa** [comment]
OK, let’s wait for the next release.

**Comment by jliljebl** [comment]
I dropped all *mltfilters.MultiFilterObject* filters and replaced them with equivalent normal filters. This fixed the Issue here since copying keyframes works for normal filters.

**Comment by jliljebl** [comment]
Yep, thanks, I pushed a patch for this just now.

**Comment by jliljebl** [comment]
Yep, I pushed a patch for this just now, please test again.

**Comment by jliljebl** [comment]
Yep, I pushed a patch for this, please re-test.

Note that OLD Volume and Brightness filters still don't work with Keyframe tool and won't be made to work. I'll add a patch that destroys them if kftool edit is attempted on them, with info window obviously.

**Comment by jep-fa** [comment]
Thanks.
Sorry it doesn't work.

When I right click on a clip -> Edit -> **Brightness Keyframes**, the keyframes curve is shown. But when I try to see the curve of the next clip the panel is empty. I get the error KeyError: 'editable_property'.

Traceback (most recent call last):
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/cairoarea.py", line 80, in _draw_event
    self._draw_func(None, cr, (a.x, a.y, a.width, a.height)) # 'None' is event object that was used to pass through here.
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/tlinewidgets.py", line 1684, in _draw
    self.edit_mode_overlay_draw_func(cr, self.edit_mode_data)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/tlinewidgets.py", line 1252, in draw_kftool_overlay
    draw_function(cr)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/kftoolmode.py", line 374, in _tline_overlay
    _kf_editor.draw(cr)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/kftoolmode.py", line 494, in draw
    self._draw_value_lines(cr, x, w)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/kftoolmode.py", line 694, in _draw_value_lines
    y = self._get_panel_y_for_value(0.0)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/kftoolmode.py", line 1229, in _get_panel_y_for_value
    editable_property = edit_data["editable_property"] 
KeyError: 'editable_property'

**With Volume Keyframes** no curve is shown and the program hangs.

  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/edit.py", line 1850, in _add_multipart_filter_redo
    self.filter_object.attach_filters(self.clip)
AttributeError: 'EditAction' object has no attribute 'filter_object'
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/cairoarea.py", line 92, in _button_press_event
    self.press_func(event)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/tlinewidgets.py", line 1485, in _press_event
    self.press_listener(event, get_frame(event.x))
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/editevent.py", line 376, in tline_canvas_mouse_pressed
    press_func(event, frame)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/kftoolmode.py", line 311, in mouse_press
    init_tool_for_clip(clip, track)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/kftoolmode.py", line 155, in init_tool_for_clip
    action.do_edit()
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/edit.py", line 339, in do_edit
    self.redo()
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/edit.py", line 409, in redo
    self.redo_func(self)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/edit.py", line 1853, in _add_multipart_filter_redo
    self.filter_object = current_sequence().create_multipart_filter(self.filter_info, self.clip)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/sequence.py", line 570, in create_multipart_filter
    filter_object = mltfilters.MultipartFilterObject(filter_info)
  File "/home/jpf/flb_git/flowblade/flowblade-trunk/Flowblade/mltfilters.py", line 286, in __init__
    self.value = copy.deepcopy(filter_info.multipart_value)
AttributeError: 'FilterInfo' object has no attribute 'multipart_value'

  In hope that helps you.

**Comment by jep-fa** [comment]
With the new patch there are bugs : Volume is OK. Brightness curve has disappeared but Brightness Level is OK. When you go to previous or next clip, the volume curve is shown even if you are in brightness.

What do you mean by 
_Note that OLD Volume and Brightness filters still don't work with Keyframe tool and won't be made to work. I'll add a patch that destroys them if kftool edit is attempted on them, with info window obviously._

The projects created with previous versions of Flowblade cannot be modified ?

**Comment by jliljebl** [comment]
> Brightness curve has disappeared but Brightness Level is OK. 

So far 
* *Brightness Keyframed* filter has been removed because it is a multipart filter and has to go now.
* Old *Brightness* didn't have keyframes and is now also removed.
* There is now only one new *Brightness* filter and that is keyframable and works with Keyframe tool
* Old  *Brightness Keyframed* and old *Brightness* don't work with Keyframe Tool as that is now made to work with the new filter

I will add code to convert old  *Brightness Keyframed* and old *Brightness* to new *Brightness* as this should be quite easy.

> When you go to previous or next clip, the volume curve is shown even if you are in brightness.

Yes, this is the current behavior, video clips are opened into Volume editing by default. Maybe this should change but that would be special casing for  *Brightness*, I don't know.

> The projects created with previous versions of Flowblade cannot be modified ?

No. 

The old *Volume* filter had value range 0-1, new  *Volume* filter has value range -70  - +20. As long as the filters are not edited they work as before but unless I find the completely correct mapping from old "0-1"T to new "-70  - +20" I have to: 
* A) destroy the old Volume filters because they cannot be correctly converted to new Volume filters.
* B) inform user that old filters cannot be edited with "Keyframe tool*, they can only be edited with Filter panel editor and tell the to replace them with new ones if user wishes to use Keyframe editor.

B) is probably better. 

I'll re-label with *In Feature queue* as this needs more work. I'll comment more once this is ready for more testing, I will probably do something else next.

**Comment by jep-fa** [comment]
Thanks for your explanations.  I agree the B) solution.

Version 2.6.1
kfpanel is empty.

Right click on a clip, popup menu, left click on Edit, left click on Volume Keyframes
Ok
left click on the hamburger menu, left click on Brightness keyframes
kfpanel is empty and several errors are shown

first error
Traceback (most recent call last):
  File "/home/jpf/flb_git_test/flowblade/flowblade-trunk/Flowblade/kftoolmode.py", line 1584, in _oor_menu_item_activated
    init_tool_for_clip(edit_data["clip"] , edit_data["track"], BRIGHTNESS_KF_EDIT)
  File "/home/jpf/flb_git_test/flowblade/flowblade-trunk/Flowblade/kftoolmode.py", line 179, in init_tool_for_clip
    _kf_editor = TLineKeyFrameEditor(ep, True, BRIGHTNESS_KF_EDIT)
  File "/home/jpf/flb_git_test/flowblade/flowblade-trunk/Flowblade/kftoolmode.py", line 388, in __init__
    self.clip_length = editable_property.get_clip_length() - 1
AttributeError: 'NoneType' object has no attribute 'get_clip_length'

It is the same behavior if you start with Brightness. Brightness is OK but now left click on the hamburger menu and the volume panel is empty and there are errors.

I don't find why.

**Comment by jep-fa** [comment]
It seems I found the reason. In kftoolmode.py, def _get_param_editable_property_with_filter_search()
the last **return None** has to have an effect after the end of the for and to be in line with the for. So I moved it back at position 4 in the line.
The hamburger menu works.

**Comment by jliljebl** [comment]
Thanks for additional info, I will circle back around to this next week and start pushing additional patches.

**Comment by jliljebl** [comment]
OK, we have disabled editing old Volume filters with Keyframe Tool and were able to open 2.6.3 created files with multipart filters successfully.

Copy pasting keyframes using the *new* Volume filter works now. This is as good as it is going to get now. Please open new Issue one if there are some further problems.
